packages/remirror__extension-file/src/file-extension.tsx

```diff
@@ -0,0 +1,375 @@
+        this.options.onDeleteFile({ tr, pos });
+        dispatch?.(tr);
+        return true;
+      }
+
+      return false;
+    };
+  }
+
+  @command()
+  renameFile(pos: number, fileName: string): CommandFunction {
+    return ({ tr, state, dispatch }) => {
+      const node = state.doc.nodeAt(pos);
+
+      if (node && node.type === this.type) {
+        dispatch?.(tr.setNodeMarkup(pos, undefined, { ...node.attrs, fileName }));
+        return true;
+      }
+
+      return false;
+    };
+  }
+
+  private preLoadFile(id: FileId, file: File, pos: number | undefined): Promise<FileLoader> | void {
+    const tr = this.store.view.state.tr;
+    const insertPos = insertFilePoint(tr.doc, isNumber(pos) ? pos : tr.selection.from, this.type);
+
+    if (!isNumber(insertPos)) {
+      // failed to find a postition to insert the file node
+      return;
+    }
+
+    // insert the file node
+    const attrs: FileAttributes = { ...this.options.getAttrsBeforeUpload(file), id };
+    tr.insert(insertPos, this.type.createChecked(attrs));
+
+    // create a fileLoader, which will read the file
+    const fileLoader: FileLoader = this.options.loadFileHandler(file);
+
+    // insert the placeholder decoration
+    setPlaceholderAction(tr, { add: { id, pos: insertPos, playload: fileLoader } });
+
+    this.store.view.dispatch(tr);
+
+    // return a promise that resolves when the file is loaded, whether
+    // successfully or unsuccessfully
+    return new Promise<FileLoader>((resolve) => {
+      fileLoader.addEventListener('loadend', () => {
+        resolve(fileLoader);
+      });
+    });
+  }
+
+  private postLoadFile(id: unknown, file: File, fileLoader: FileLoader) {
+    const view = this.store.view;
+    const placeholderPos = findPlaceholderPos(view.state, id);
+
+    // unexpected
+    if (placeholderPos == null) {
+      return;
+    }
+
+    const $pos = view.state.doc.resolve(placeholderPos);
+    const fileNode = $pos.nodeAfter;
+
+    // if the file node around the placeholder has been deleted, then delete
+    // the placeholder and drop the uploaded file.
+    if (!fileNode) {
+      const tr = view.state.tr;
+      setPlaceholderAction(tr, { remove: { id } });
+      view.dispatch(tr);
+      return;
+    }
+
+    // unexpected
+    if (fileNode.type !== this.type) {
+      return;
+    }
+
+    // unexpected
+    if (fileNode.attrs.id !== id) {
+      return;
+    }
+
+    // Update the file node at the placeholder's position, and remove
+    // the placeholder.
+    const tr = view.state.tr;
+    setPlaceholderAction(tr, { remove: { id } });
+    const attrs: FileAttributes = {
+      ...fileNode.attrs,
+      ...this.options.getAttrsAfterUpload(file, fileLoader),
+      id: null,
+    };
+    tr.setNodeMarkup(placeholderPos, undefined, attrs);
+    view.dispatch(tr);
+  }
+
+  private startLoadFile(file: File, pos?: number | undefined): void {
+    // A fresh object to act as the ID for this upload.
+    const id = {};
```

Comment:
I stole this line of code and its comment from the official example: https://prosemirror.net/examples/upload/

When we use an empty object as an ID, we actually use its memory address. This brings two advantages over using a unique string as an id:

1. Eliminate the possibility of duplicate IDs (Although the probability is very small)
2. Increase performance since we don't need a complex algorithm to generate a string